runtime/compiler/optimizer/JProfilingRecompLoopTest.hpp

```diff
@@ -0,0 +1,57 @@
+ * Copyright (c) 2018, 2018 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+#include <stdint.h>                           // for int32_t
+#include "optimizer/Optimization.hpp"         // for Optimization
+#include "optimizer/OptimizationManager.hpp"  // for OptimizationManager
+#include "infra/Checklist.hpp"                // fir NodeChecklist
+#include "runtime/J9Profiler.hpp"
+#include "optimizer/Structure.hpp"
+/**
+ * Class TR_JProfilingRecompLoopTest
+ * ===================================
+ *
+ * When jProfiling in compilations are enabled, we have only one recompilation
+ * test that counts the method invocation to trip up recompilation. For work loads
+ * where we are spending a lot of time in loops, we are hitting a risk where we
+ * are running into compiled method body with profiling trees if we just rely on
+ * method invocations. This optimization scans the method body and puts 
+ * recompilation test after asynccheck node to consider number of times a loop
+ * is running.
+ */
+class TR_JProfilingRecompLoopTest : public TR::Optimization
+   {
+   public:
+   typedef TR::typed_allocator<std::pair<std::pair<TR::TreeTop*, TR::Block*>, int32_t>, TR::Region &> RecompilationTestLocationAllocator;
+   typedef std::list<std::pair<std::pair<TR::TreeTop*, TR::Block*>, int32_t>, RecompilationTestLocationAllocator> RecompilationTestLocations;
+   TR_JProfilingRecompLoopTest(TR::OptimizationManager *manager)
+      : TR::Optimization(manager)
+      {}
+   static TR::Optimization *create(TR::OptimizationManager *manager)
+      {
+      return new (manager->allocator()) TR_JProfilingRecompLoopTest(manager);
+      }
+   virtual int32_t perform();
+   virtual const char *optDetailString() const throw();
+   void addRecompilationTests(TR::Compilation *comp, RecompilationTestLocations &testLocations);
+   int32_t getLoopNestingDepth(TR::Compilation *comp, TR::Block *block);
```

Comment:
Isn't the structure's getNestingDepth what you want to use for this? At the very lest if this doesn't yet exist (and it should) then it would need to be pulled off the opt and put somewhere in structure as a utility.

Reply:
I need to look in depth how Structure gets nesting depth. I walk parent structure till the point I find acyclic structure.

Reply:
I moved this utility function into the `J9TransformUtil` in https://github.com/eclipse/openj9/pull/2348/commits/aa1b8bc31d69fd0268cb59a8d69455505ba8ec2c